packages/protocol/contracts/layer1/based/TaikoInbox.sol

```diff
@@ -0,0 +1,723 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "src/shared/common/EssentialContract.sol";
+import "src/shared/based/ITaiko.sol";
+import "src/shared/libs/LibAddress.sol";
+import "src/shared/libs/LibMath.sol";
+import "src/shared/libs/LibNetwork.sol";
+import "src/shared/libs/LibStrings.sol";
+import "src/shared/signal/ISignalService.sol";
+import "src/layer1/verifiers/IVerifier.sol";
+import "./IFork.sol";
+import "./ITaikoInbox.sol";
+
+// import "forge-std/src/console2.sol";
+
+/// @title TaikoInbox
+/// @notice Acts as the inbox for the Taiko Alethia protocol, a simplified version of the
+/// original Taiko-Based Contestable Rollup (BCR). The tier-based proof system and
+/// contestation mechanisms have been removed.
+///
+/// Key assumptions of this protocol:
+/// - Block proposals and proofs are asynchronous. Proofs are not available at proposal time,
+///   unlike Taiko Gwyneth, which assumes synchronous composability.
+/// - Proofs are presumed error-free and thoroughly validated, with proof type management
+///   delegated to IVerifier contracts.
+///
+/// @dev Registered in the address resolver as "taiko".
+/// @custom:security-contact security@taiko.xyz
+abstract contract TaikoInbox is EssentialContract, ITaikoInbox, ITaiko, IFork {
+    using LibMath for uint256;
+
+    State public state; // storage layout much match Ontake fork
+
+    // External functions ------------------------------------------------------------------------
+
+    function init(
+        address _owner,
+        address _rollupResolver,
+        bytes32 _genesisBlockHash
+    )
+        external
+        initializer
+    {
+        __Taiko_init(_owner, _rollupResolver, _genesisBlockHash);
+    }
+
+    /// @notice Proposes a batch of blocks.
+    /// @param _params ABI-encoded parameters consisting of:
+    /// - proposer: The address of the proposer, which is set by the PreconfTaskManager if
+    ///             enabled; otherwise, it must be address(0).
+    /// - coinbase: The address that will receive the block rewards; defaults to the proposer's
+    ///             address if set to address(0).
+    /// - batchParams: Batch parameters.
+    /// @param _txList      The transaction list in calldata.
+    /// @return meta_       Batch metadata.
+    function proposeBatch(
+        bytes calldata _params,
+        bytes calldata _txList
+    )
+        external
+        nonReentrant
+        returns (BatchMetadata memory meta_)
+    {
+        Stats2 memory stats2 = state.stats2;
+        require(!stats2.paused, ContractPaused());
+
+        Config memory config = getConfig();
+        require(stats2.numBatches >= config.forkHeights.pacaya, InvalidForkHeight());
+
+        unchecked {
+            require(
+                stats2.numBatches < stats2.lastVerifiedBatchId + config.maxBatchProposals,
```

Comment:
ditto

Reply:
fixed in https://github.com/taikoxyz/taiko-mono/pull/18800